src/compiler/checker.ts

```diff
@@ -4438,15 +4510,15 @@ namespace ts {
                 let reportStructuralErrors = reportErrors && errorInfo === saveErrorInfo;
                 // identity relation does not use apparent type
                 let sourceOrApparentType = relation === identityRelation ? source : getApparentType(source);
-                if (sourceOrApparentType.flags & TypeFlags.ObjectType && target.flags & TypeFlags.ObjectType) {
+                if (sourceOrApparentType.flags & (TypeFlags.ObjectType | TypeFlags.Intersection) && target.flags & TypeFlags.ObjectType) {

```

Comment:
This part is very interesting. Can you put a comment explaining why we only do this for intersections, and only on the source side? I do understand why, it's because we want the intersection to be equivalent to the meet (the aggregation of all the properties), but the union type is not similarly equivalent to the join (the properties the two types have in common). And it's only on the source side because the intent of this step is widen assignability, to allow more things than the union/intersection steps allowed.